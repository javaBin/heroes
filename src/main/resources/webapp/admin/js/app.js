"use strict";

var heroes = [
  {navn: "Jaran Flaath", bidrag: "Nestleder", email: "j.f@g.com", fraDato: "2015-01-01"},
  {navn: "Anders Karlsen", bidrag: "Aktiv", email: "a.k@g.com", fraDato: "2015-01-01"},
  {navn: "Bjørn Hamre", bidrag: "Styremedlem", email: "b.h@g.com", fraDato: "2015-01-01"}
];

var Button = ReactBootstrap.Button;
var Input = ReactBootstrap.Input;
var Grid = ReactBootstrap.Grid;
var Row = ReactBootstrap.Row;
var Col = ReactBootstrap.Col;
var Table = ReactBootstrap.Table;

var NewHeroForm = React.createClass({
  render: function () {
    return (
      <div className="heroes-admin-add">
        <Grid>
          <Row>
            <Col md={6}>
              <Input type="text" label="Navn" placeholder="Heltens navn"/>
            </Col>
            <Col md={6}>
              <Input type="email" label="Email" placeholder="hero@java.no"/>
            </Col>
          </Row>
          <Row>
            <Col md={6}>
              <Input type="select" label="Heltetype">
                <option value="styremedlem">Styremedlem</option>
                <option value="foredragsholder-jz">Foredragsholder på JavaZone</option>
                <option value="foredragsholder">Foredragsholder på javaBin</option>
                <option value="regionsleder">Regionsleder</option>
                <option value="aktiv">Aktiv</option>
              </Input>
            </Col>
            <Col md={6}>
              <Input type="date" label="Fra dato"/>
            </Col>
          </Row>
        </Grid>
        <Button bsStyle="info">Legg til</Button>
      </div>
    );
  }
});


var HeroesList = React.createClass({
  render: function () {

    var heroComponents = this.props.data.map(function (hero) {
      return (
        <tr>
          <td>
            <img height="25" src="images/hero.png"/>
          </td>
          <td>
            {hero.navn}
          </td>
          <td>
            {hero.bidrag}
          </td>
        </tr>
      );
    });
    return (
      <div className="heroes-list">
        <h2>Helter</h2>
        <Table striped bordered condensed hover>
          <thead>
            <tr>
              <th></th>
              <th>Navn</th>
              <th>Heltetype</th>
            </tr>
          </thead>
          <tbody>
            {heroComponents}
          </tbody>
        </Table>
      </div>
    );
  }
});

var HeroesAdmin = React.createClass({
  render: function () {
    return (
      <div className="heroes-admin-container">
        <Grid>
          <Row>
            <Col md={12}>
              <h1>javaBin Heroes</h1>
              <NewHeroForm />
              <HeroesList data={this.props.data}/>
            </Col>
          </Row>
        </Grid>
      </div>
    );

  }
});
ReactDOM.render(<HeroesAdmin data={heroes}/>, document.getElementById("heroesadmin"));
