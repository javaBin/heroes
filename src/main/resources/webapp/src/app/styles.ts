import { createStyles, Theme } from "@material-ui/core";

export const styles = ({ spacing }: Theme) =>
  createStyles({
    heroesList: {},
    paper: {
      padding: spacing.unit,
      margin: spacing.unit * 3
    }
  });
