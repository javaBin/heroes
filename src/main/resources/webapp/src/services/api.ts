export interface Person {
  name: string;
  email: string;
  twitter?: string;
  avatar_image?: string;
}

export interface HeroAchievement {
  id?: string;
  type: Achievement;
  label?: string;
}

export interface Hero extends Person {
  achievements: HeroAchievementDetail[];
  id?: string;
  published?: boolean;
}

export interface Userinfo {
  authenticated?: boolean;
  admin?: boolean;
  username?: string;
}

export type Achievement = "FOREDRAGSHOLDER_JZ" | "FOREDRAGSHOLDER_JAVABIN" | "STYRE";

export function allAchievements(): Achievement[] {
  return ["FOREDRAGSHOLDER_JZ", "FOREDRAGSHOLDER_JAVABIN", "STYRE"];
}

export function achievementName(achievement: Achievement): string {
  switch (achievement) {
    case "FOREDRAGSHOLDER_JAVABIN":
      return "Foredragsholder JavaBin";
    case "FOREDRAGSHOLDER_JZ":
      return "JavaZone foredragsholder";
    case "STYRE":
      return "Styre";
  }
}

export interface CreateHeroData {
  people: Person[];
}

export interface ConferenceSpeakerAchievement extends HeroAchievement {
  type: "FOREDRAGSHOLDER_JZ";
  year: number;
  title: string;
}

export interface UsergroupSpeakerAchievement extends HeroAchievement {
  type: "FOREDRAGSHOLDER_JAVABIN";
  date: string;
  title: string;
}

export type BoardMemberRole = "BOARD_MEMBER" | "VICE_CHAIR" | "CHAIR";

export function boardMemberRoleName(role: BoardMemberRole): string {
  switch (role) {
    case "BOARD_MEMBER":
      return "Board member";
    case "CHAIR":
      return "Chair";
    case "VICE_CHAIR":
      return "Vice chair";
  }
}

export interface BoardMemberAchivement extends HeroAchievement {
  type: "STYRE";
  year: number;
  role: BoardMemberRole;
}

export type HeroAchievementDetail = ConferenceSpeakerAchievement | UsergroupSpeakerAchievement | BoardMemberAchivement;

interface Consent {
  id: number;
  text: string;
}

export interface HeroProfile {
  profile: Person;
  published?: boolean;
  consent?: Consent;
  achievements: HeroAchievement[];
}

export interface HeroService {
  deleteAchievement(heroId: string, achievementId: string): Promise<void>;
  updateAchievement(heroId: string, achievementId: string, achievement: HeroAchievementDetail): Promise<void>;
  addAchievement(heroId: string, achievement: HeroAchievementDetail): Promise<void>;
  updateHero(heroId: string, update: Partial<Hero>): Promise<void>;
  fetchCreateHeroData(): Promise<CreateHeroData>;
  fetchUserinfo(): Promise<Userinfo>;
  fetchMe(): Promise<HeroProfile>;
  fetchHeroDetails(heroId: string): Promise<Hero>;
  fetchPublicHeroes(): Promise<Hero[]>;
  fetchHeroes(): Promise<Hero[]>;
  addHero(hero: Hero): Promise<void>;
  consentToPublish(consentId: number): Promise<void>;
}
