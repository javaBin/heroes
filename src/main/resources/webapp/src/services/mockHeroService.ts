import { CreateHeroData, Hero, HeroService } from "./heroService";

interface User {
    admin: boolean;
    email: string;
}

export class MockHeroService implements HeroService {
    currentUser?: User = {
        admin: true,
        email: "johannes@brodwall.com",
    };

    private heroes: Hero[] = [];

    async fetchCreateHeroData() {
        return {
            achievements: [],
            people: [],
        };
    }
    async consentToPublish() {
        this.heroes.find(h => h.email === this.currentUser!.email)!.published = true;
    }

    async fetchUserinfo() {
        return {
            authenticated: !!this.currentUser,
            username: this.currentUser && this.currentUser.email,
        };
    }

    async fetchMe() {
        return this.heroes.find(h => h.email === this.currentUser!.email)!;
    }
    async addHero(hero: Hero) {
        this.heroes.push(hero);
    }
    async fetchHeroes() {
        return this.heroes.filter(h => h.published);
    }

}
