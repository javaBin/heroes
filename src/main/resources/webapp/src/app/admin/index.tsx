import React, { FormEvent } from "react";
import { HeroService } from "../../services";
import { Achievement, achievementName, allAchievements, CreateHeroData, Hero } from "../../services/heroService";

import heroPng from "../../images/hero.png";
import { HeroControlPanel } from "./HeroControlPanel";

interface AdminProps {
  heroService: HeroService;
}

interface AdminState {
  createHeroData?: CreateHeroData;
  heroes?: Hero[];
}

export class AdminScreen extends React.Component<AdminProps, AdminState> {
  state: AdminState;

  constructor(props: AdminProps) {
    super(props);
    this.state = {};
  }

  async componentDidMount() {
    const {heroService} = this.props;
    const [createHeroData, heroes] = [await heroService.fetchCreateHeroData(), await heroService.fetchHeroes()];
    this.setState({ createHeroData, heroes });
  }

  handleNewHero = async (hero: Hero) => {
    await this.props.heroService.addHero(hero);
    const {heroService} = this.props;
    const [createHeroData, heroes] = [await heroService.fetchCreateHeroData(), await heroService.fetchHeroes()];
    this.setState({ createHeroData, heroes });
  }

  render() {
    const {createHeroData, heroes} = this.state;
    if (!createHeroData || !heroes) {
      return <div>Loading...</div>;
    }
    return <>
      <div className="heroes-admin-container">
        <h1>javaBin Heroes</h1>
        <h2>Add Hero</h2>
        <NewHeroForm onNewHero={this.handleNewHero} createHeroData={createHeroData} />
        <h2>Current heroes</h2>
        <AdminHeroList heroes={heroes} />
      </div>
      <fieldset>
        <h2>UX Proof-of-concept heroes control panel:</h2>
        <HeroControlPanel heroes={heroes} people={createHeroData.people} prefix="#admin" />
      </fieldset>
    </>;

  }
}

function AdminHeroList({heroes}: {heroes: Hero[]}) {
  const heroComponents = heroes.map((hero) => {
    return (
      <tr key={hero.id}>
        <td>
          <img height="25" src={heroPng} />
        </td>
        <td>
          {hero.name}
        </td>
        <td>
          {hero.achievement}
        </td>
        <td>
          {hero.published ? "Published" : "Unpublished"}
        </td>
      </tr>
    );
});
  return (
    <div className="heroes-list">
      <h2>Helter</h2>
      <table>
        <thead>
          <tr>
            <th></th>
            <th>Navn</th>
            <th>Heltetype</th>
            <th>Published</th>
          </tr>
        </thead>
        <tbody>
          {heroComponents}
        </tbody>
      </table>
    </div>
  );
}

function FormControl({label, type, value, onChange}:
  {label: string, type?: string, value: string, onChange: (s: string) => void},
  ) {
  return <label>
    {label}
    <input type={type} value={value} onChange={event => onChange(event.target.value)} />
  </label>;
}

function DateControl({label}: {label: string}) {
  return <label>
    {label}
    <input type="date" />
  </label>;
}

interface SelectOption { value: string; label: string; }

function SelectControl({label, options, value, onChange, includeBlank}:
  {label: string, options: SelectOption[], value: string, onChange: (s: string) => void, includeBlank: boolean},
  ) {
  return <label>
    {label}
    <select value={value} onChange={event => onChange(event.target.value)}>
      {includeBlank && <option></option>}
      {options.map(o => <option key={o.value} value={o.value}>{o.label}</option>)}
    </select>
  </label>;
}

class NewHeroForm extends React.Component<{onNewHero: (h: Hero) => void, createHeroData: CreateHeroData}> {
    state = {
      achievement: "",
      email: "",
      name: "",
    };

    handleSubmit = async (e: FormEvent) => {
      const {email, name, achievement} = this.state;
      const newHero = {
        achievement, achievements: [], avatar: "", email, name,
        published: false,
      };
      e.preventDefault();
      await this.props.onNewHero(newHero);
      this.setState({achievement: "", email: "", name: ""});
    }

    render = () => {

      const {name, email, achievement} = this.state;

      const people = this.props.createHeroData.people.map(p => ({label: p.name, value: p.email}));
      const achievements = allAchievements();

      return (
        <div className="heroes-admin-add">
          <form onSubmit={this.handleSubmit}>
            <SelectControl
              label="Person"
              options={people}
              value={email}
              onChange={(email) => this.setState({email})}
              includeBlank={true}
            />
            <FormControl label="Hero name" value={name} onChange={(name) => this.setState({name})} />
            <FormControl label="Hero email" type="email" value={email} onChange={(email) => this.setState({email})} />
            <SelectControl
              label="Hero type"
              options={achievements.map(a => ({value: Achievement[a], label: achievementName(a)}))}
              value={achievement}
              onChange={(achievement) => this.setState({achievement})}
              includeBlank={true}
            />
            <DateControl label="Dato" />
            <button>Legg til</button>
          </form>
        </div>
      );
    }
}
