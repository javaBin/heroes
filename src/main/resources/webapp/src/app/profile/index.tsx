import React from "react";
import { HeroProfile, HeroService } from "../../services";

interface ProfileState {
    loaded: boolean;
    profile: HeroProfile;
}

export class ProfileScreen extends React.Component<{heroService: HeroService}, ProfileState> {
    state = {
        loaded: false,
        profile: {
            consent: { id: 0, text: "" },
            heroism: { achievement: "" },
            profile: { name: "", email: "" },
        },
    };

    componentDidMount = async () => {
        const profile = await this.props.heroService.fetchMe();
        this.setState({profile, loaded: true});
    }

    handlePublishSubmit = async () => {
        await this.props.heroService.consentToPublish(this.state.profile.consent.id);
        const profile = await this.props.heroService.fetchMe();
        this.setState({profile, loaded: true});
    }

    render() {
        if (!this.state.loaded) {
            return <div>Loading...</div>;
        }
        const {profile} = this.state;
        if (!profile.heroism) {
            return <div>You are not a hero, {profile.profile.name}</div>;
        }
        return <>
            <h1>{profile.profile.name}</h1>
            <div>{profile.heroism.achievement}</div>
            {profile.consent && <>
                <h2>We need your consent to publish the information about you</h2>
                {profile.consent.text}
                <button onClick={this.handlePublishSubmit}>I agree to be published</button>
            </>}
        </>;
    }
}
