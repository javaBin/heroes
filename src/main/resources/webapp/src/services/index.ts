import { HeroProfile, HeroService } from "./api";

export { HeroProfile, HeroService };
