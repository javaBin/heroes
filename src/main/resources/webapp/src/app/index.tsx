import React from "react";
import { HeroService } from "../services";
import { Hero, Userinfo } from "../services/api";
import { AdminScreen } from "./admin";

import { Table } from "react-bootstrap";
import { ProfileScreen } from "./profile";

import heroPng from "../images/hero.png";
import { HeroServiceHttp } from "../services/heroServiceHttp";
import { HeroControlPanel } from "./admin/HeroControlPanel";

export function HeroList({heroes}: {heroes: Hero[]}) {
    const heroComponents = heroes.map((hero) => {
        return (
          <tr key={hero.email}>
            <td>
              <img height="25" src={heroPng} />
            </td>
            <td>
              {hero.name}
            </td>
            <td>
              {hero.achievement}
            </td>
          </tr>
        );
    });
    return (
        <div className="heroes-list">
          <h2>Helter</h2>
          <Table striped bordered condensed hover>
            <thead>
              <tr>
                <th></th>
                <th>Navn</th>
                <th>Heltetype</th>
              </tr>
            </thead>
            <tbody>
              {heroComponents}
            </tbody>
          </Table>
        </div>
      );
}

export class HeroListComponent extends React.Component<{heroService: HeroService}, {heroes?: Hero[], loaded: boolean}> {
    state = {
        heroes: undefined,
        loaded: false,
    };

    componentDidMount = async () => {
        const heroes = await this.props.heroService.fetchHeroes();
        this.setState({heroes, loaded: true});
    }

    render() {
        if (!this.state.loaded) {
            return <div>Loading....</div>;
        }
        return <HeroList heroes={this.state.heroes!} />;
    }
}

export class App extends React.Component<{heroService: HeroService}, {hash: string, userinfo: Userinfo}> {

    state = {
        hash: window.location!.hash,
        userinfo: { authenticated: false, username: "", admin: false },
    };

    handleHashchange = () => {
        const {hash} = window.location;
        this.setState({hash});
    }

    async componentDidMount() {
        window.addEventListener("hashchange", this.handleHashchange);
        const userinfo = await this.props.heroService.fetchUserinfo();
        this.setState({userinfo});
    }

    componentWillUnmount() {
        window.removeEventListener("hashchange", this.handleHashchange);
    }

    renderContent() {
        const { hash } = this.state;
        if (hash.indexOf("#admin") === 0) {
            return <AdminScreen heroService={this.props.heroService} />;
        }

        if (hash.indexOf("#profile") === 0) {
            return <ProfileScreen heroService={this.props.heroService} />;
        }

        if (hash.indexOf("#poc") === 0) {
            return <HeroControlPanel heroService={new HeroServiceHttp()} prefix="#poc" />;
        }

        return (
            <div>
                <h1>{hash}</h1>
                <HeroListComponent heroService={this.props.heroService} />
            </div>
        );
    }

    render() {
        const {userinfo} = this.state;
        return <>
            <nav>
                <ul>
                    <li><a href="#">View heroes</a></li>
                    {userinfo.authenticated || <li><a href="/api/login">Log in</a></li>}
                    {userinfo.admin && <li><a href="#admin">Admin</a></li>}
                    {userinfo.authenticated && <li><a href="#profile">{userinfo.username}</a></li>}
                </ul>
            </nav>
            {this.renderContent()}
        </>;
    }

}
