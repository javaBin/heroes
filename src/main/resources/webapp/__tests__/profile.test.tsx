import React from "react";
import renderer from "react-test-renderer";
import { ProfileScreen } from "../src/app/profile";
import { MockHeroService } from "../src/services/mockHeroService";

function promiseCompletion() {
    return new Promise((resolve) => setImmediate(resolve));
}

describe("profile screen", () => {

    it("shows profile screen", async () => {
        const app = renderer.create(<ProfileScreen heroService={new MockHeroService()} />);
        await promiseCompletion();
        expect(app.toJSON()).toMatchSnapshot();
    });

});
