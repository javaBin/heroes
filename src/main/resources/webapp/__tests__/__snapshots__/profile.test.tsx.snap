// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile screen shows profile screen 1`] = `
<div>
  You are not a hero, 
  Johannes
</div>
`;
